File: Data Structures & Algorithms/reverse-a-linked-list/submission-8.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func reverseList(head *ListNode) *ListNode {
    //set all pointers
    curr := head
	var prev *ListNode = nil
    var next *ListNode = nil

	/*
	   A -> B -> C
	   next = curr.next
	   curr.next = prev
	   prev = curr
	   curr = next
	*/
	for curr != nil{
        next = curr.Next
		curr.Next = prev
		prev = curr
		curr = next
	}

	return prev
}

File: Data Structures & Algorithms/linked-list-cycle-detection/submission-8.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    #use fast poitner that goes nex.next and slow ptr that jsut gos next
    #the place where fast point adn slow point inercet is where cycle is
    #if they both exit, not cycle
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        slowPtr = head
        fastPtr = head

        while fastPtr and fastPtr.next:
            fastPtr = fastPtr.next.next
            slowPtr = slowPtr.next
            if slowPtr == fastPtr:
               return True 

        return False